File: Data Structures & Algorithms/binary-tree-from-preorder-and-inorder-traversal/submission-2.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public TreeNode buildTree(int[] preorder, int[] inorder) {
    Map <Integer , Integer > map =new HashMap <> ();
    for(int i=0;i<inorder.length;i++){
        map.put(inorder[i],i);
    }
    TreeNode root=buildTree(preorder,0,preorder.length-1,inorder,0,inorder.length-1,map);
    return root;
    }
    public TreeNode buildTree(int[] preorder,int preSt,int preEn,int []inorder,int inSt,int inEn,Map<Integer,Integer> map){
        if(preSt>preEn || inSt >inEn) return null;
        TreeNode root=new TreeNode(preorder[preSt]);
        
        int inR=map.get(root.val);
        int numLeft=inR-inSt;
        
        root.left=buildTree(preorder,preSt+1,preSt+numLeft,inorder,inSt,inR-1,map);
        root.right=buildTree(preorder,preSt+numLeft+1,preEn,inorder,inR+1,inEn,map);
        
        return root;
    }
}

File: Data Structures & Algorithms/valid-binary-search-tree/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        l=float('-inf')
        r=float('inf')
        return BST(l,r,root)
def BST(l,r,root) -> bool:
    if root==None:
        return True
    if root.val<=l or root.val>=r:
        return False
    return BST(l,root.val,root.left) and BST(root.val,r,root.right)